### PR TITLE
Smooth shuffle streamlines

### PR DESCRIPTION
**Pull request by frheault**
Transfert of simple script from my personal repo.
Shuffle streamlines order and smoothing streamlines (using either spline or gaussian)

Testing Data:
https://drive.google.com/open?id=1RBx8717eIwx4PolY77rqTLxb4z9P0KBH

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -17,4 +18,4 @@ pytz==2018.4
 scipy==1.0.1
 setuptools==39.0.1
 six==1.11.0
-coloredlogs==10.0
+sklearn==0.0
```
Comment (AntoineTheb) [inline requirements.txt]: sklearn version 0 ?

File: scilpy/tracking/tools.py [hunk 2]
```diff
@@ -127,6 +129,53 @@ def resample_streamlines(streamlines, num_points=0, arc_length=False):
     return resampled_streamlines
 
 
+def smooth_line_gaussian(streamline, sigma):
+    if sigma < 0.00001:
+        ValueError('Cant have a 0 sigma with gaussian.')
+
+    nb_points = int(length(streamline))
+    if nb_points < 2:
+        logging.debug('Streamline shorter than 1mm, corner cases possible.')
```
Comment (AntoineTheb) [inline scilpy/tracking/tools.py]: Does this presume that the streamlines are resampled to voxmm ?

File: scripts/scil_shuffle_streamlines.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    args = parser.parse_args()
+
+    assert_inputs_exist(parser, args.in_tractogram)
+    assert_outputs_exist(parser, args, args.out_tractogram)
+
+    sft = load_tractogram_with_reference(parser, args, args.in_tractogram)
+    streamlines = list(sft.get_streamlines_copy())
+    random.shuffle(streamlines, random=args.seed)
+
+    smoothed_sft = StatefulTractogram(streamlines, sft, Space.RASMM)
```
Comment (AntoineTheb) [inline scripts/scil_shuffle_streamlines.py]: Is the smoothing part missing ?

File: requirements.txt (hunk 1 above)
Comment (frheault) [reply requirements.txt]: I know, that's from my pip freeze (after an upgrade).

File: scilpy/tracking/tools.py (hunk 2 above)
Comment (frheault) [reply scilpy/tracking/tools.py]: Yesm either voxmm or rasmm

File: scripts/scil_shuffle_streamlines.py (hunk 3 above)
Comment (frheault) [reply scripts/scil_shuffle_streamlines.py]: That would be in the other script, scil_smooth_streamlines.py

Comment (AntoineTheb) [reply scripts/scil_shuffle_streamlines.py]: It really is kind of nitpicky from my part and somewhat unnecessary, but since the tractogram is not smoothed, only shuffled in this script, could the variable be renamed to something lilke `shuffled_sft` to avoid potential confusion ?

Comment (frheault) [reply scripts/scil_shuffle_streamlines.py]: Ahh ok, I didn't understand the question. I thought you said that because the name of the PR start with smooth and you were confused.

And I was the one confuse, 100% my bad. Sorry, you are right.

File: requirements.txt (hunk 1 above)
Comment (frheault) [reply requirements.txt]: Should I trust pip freeze ? I don't know what version to write

Comment (AntoineTheb) [reply requirements.txt]: From a brief google search, I found this:
https://stackoverflow.com/questions/46991795/why-does-pip-freeze-geneate-package-name-0-0-0

which, according to the question, makes the package installation break. I would guess it is safer to remove this line

Comment (AntoineTheb) [reply requirements.txt]: After 30 seconds more of research, I came across this:
https://github.com/scikit-learn/scikit-learn/issues/8215
This does provide more information, but no answers

Comment (frheault) [reply requirements.txt]: Actually it was a wrong commit from another PR. I removed it.
But I will use scikit-learn==0.22.1 now (I think it is an alias problem)

File: scripts/scil_smooth_streamlines.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+    This script will smooth the streamlines, usually to remove the
+    'wiggles' in probabilist tracking, two choice of method are available:
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: probabilistIC
choiceS of methodS

File: scripts/scil_smooth_streamlines.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        - Gaussian will use the surrounding coordinates for smoothing.
+        Streamline are resampled to 1mm step-size and the smoothing is
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: StreamlineS

File: scripts/scil_smooth_streamlines.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        performed on the coordinate array. The sigma will be indicative of the
+        number of points surrounding the center points to be use for blurring.
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: useD

File: scripts/scil_smooth_streamlines.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+        - Spline will fit a sline curve to every streamlines using a sigma and
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: sline -> spline
streamlines -> streamline

File: scripts/scil_smooth_streamlines.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        the number of control points. The sigma represent the allowed distance
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: representS

File: scripts/scil_smooth_streamlines.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        from the control points. The control points for the spline fit will be
+        the resampled streamline.
+        WARNING: Too low of a sigma (e.g: 1) with a lot of control
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: Add newline before WARNING please

File: scripts/scil_smooth_streamlines.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        points (e.g: 15) will create crazy streamlines that could end up of the
+        bounding box.
+
+        This script enforce endpoints to remain the same.
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: enforceS

File: scripts/scil_smooth_streamlines.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                   help='Input tractography file.')
+
+    p.add_argument('out_tractogram',
+                   help='Output tractography file.')
+
+    sub_p = p.add_mutually_exclusive_group(required=True)
+    sub_p.add_argument('--gaussian', metavar='SIGMA', type=int,
+                       help='Sigma for smoothing. Use the value of surronding\n'
+                            'X,Y,Z points on the streamline to blur the'
+                            ' streamlines.')
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: Can you suggest a good or safe sigma in the description here please

File: scripts/scil_smooth_streamlines.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    sub_p.add_argument('--spline', nargs=2, metavar=('SIGMA', 'NB_CTRL_POINT'),
+                       type=int,
+                       help='Sigma for smoothing. Model each streamline as a'
+                            'spline.')
```
Comment (mdesco) [inline scripts/scil_smooth_streamlines.py]: Again, a suggestion would be helpful

File: scripts/scil_smooth_streamlines.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+    This script will smooth the streamlines, usually to remove the
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: I would remove the indentation.

Also, a new sentence here:
`wiggles' in probabilistic tracking. Two choices of methods are available:`

And remove the additional indent for the warning paragraph.

File: scripts/scil_shuffle_streamlines.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+                             assert_inputs_exist,
+                             assert_outputs_exist)
+
+
+def _build_args_parser():
+
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Shuffle the ordering of streamlines.')
+
```
Comment (EmmaRenauld) [inline scripts/scil_shuffle_streamlines.py]: description=__doc__

File: scripts/scil_smooth_streamlines.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+        - Spline will fit a spline curve to every streamline using a sigma and
+        the number of control points. The sigma represents the allowed distance
+        from the control points. The control points for the spline fit will be
+        the resampled streamline.
+
+        WARNING: Too low of a sigma (e.g: 1) with a lot of control
+        points (e.g: 15) will create crazy streamlines that could end up of the
+        bounding box.
+        This script enforces endpoints to remain the same.
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: + WARNING. Data_per_point will be lost.

File: scripts/scil_smooth_streamlines.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+                   help='Input tractography file.')
+
+    p.add_argument('out_tractogram',
+                   help='Output tractography file.')
+
+    sub_p = p.add_mutually_exclusive_group(required=True)
+    sub_p.add_argument('--gaussian', metavar='SIGMA', type=int,
+                       help='Sigma for smoothing. Use the value of surronding\n'
+                            'X,Y,Z points on the streamline to blur the'
+                            ' streamlines.\nSigma value around 5.')
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: Not clear.

'A good sigma choice would probably be around 5.' ?

File: scripts/scil_smooth_streamlines.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    sub_p.add_argument('--spline', nargs=2, metavar=('SIGMA', 'NB_CTRL_POINT'),
+                       type=int,
+                       help='Sigma for smoothing. Model each streamline as a'
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: missing a space after "as a"

File: scripts/scil_smooth_streamlines.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+                            'spline.\nSigma value around 5 and control point '
+                            ' around 10.')
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: same comment for the "around 5"

File: scripts/scil_smooth_streamlines.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    This script will smooth the streamlines, usually to remove the
+    'wiggles' in probabilistic tracking, two choices of methods are available:
+        - Gaussian will use the surrounding coordinates for smoothing.
+        Streamlines are resampled to 1mm step-size and the smoothing is
+        performed on the coordinate array. The sigma will be indicative of the
+        number of points surrounding the center points to be used for blurring.
+
+        - Spline will fit a spline curve to every streamline using a sigma and
+        the number of control points. The sigma represents the allowed distance
+        from the control points. The control points for the spline fit will be
```
Comment (EmmaRenauld) [inline scripts/scil_smooth_streamlines.py]: the allowed distance (in mm? in vox? in the tractogram space?)
